adding all working changes for jenkins

nikhilnanal opened the pull request.

zhngaj committed:
prov/efa: Enable shared-memory support

Enable shared-memory based transfer with SHM provider for intra-node
communication.

1. For small/medium (adjustable threshold by macro RXR_MEDIUM_SIZE) message,
sender sends RTS packet with real data payload to the receiver via shm
provider. Receiver receives it and extracts the data to the user buffer.

2. For large message, sender only sends the RTS packet with IOV structures
to the receiver. Upon receiving the RTS packet, the receiver first extracts
the IOV structures. Then the receiver issues a RMA READ operation to the
sender's address (included in IOV) to get the real data, which utilizes CMA
copy. Upon receiving the completion of the RMA READ operation, the receiver
sends a EOR packet to notify the sender that data copy is done, so that
sender can write send completion.

Runtime environment variable 'FI_EFA_ENABLE_SHM_TRANSFER' can be used to
enable/disable shared memory based data transfer. It's enabled by default.

This commit needs to bump the version to be protocol compatible.

Signed-off-by: Jie Zhang <zhngaj@amazon.com>

alex-mckinley committed:
fabtests/multinode: runfabtests.sh now runs multinode test at 16, 32, 48, and 64 processes

Signed-off-by: Alex McKinley <alex.mckinley@intel.com>

alex-mckinley committed:
fabtests/multinode: changed the way address length is done due to different address sizes on OSX devices

Signed-off-by: Alex McKinley <alex.mckinley@intel.com>

alex-mckinley committed:
fabtests/multinode: bug fix: type mismatch

Signed-off-by: Alex McKinley <alex.mckinley@intel.com>

alex-mckinley committed:
fabtests/multinode: psm2 and rxd require a name server to be set up before adding clients

Signed-off-by: Alex McKinley <alex.mckinley@intel.com>

alex-mckinley committed:
fabtests/multinode: allow multinode test to take -I flag to set iterations

Signed-off-by: Alex McKinley <alex.mckinley@intel.com>

Commit:
prov/rxm: env var to force auto progress

Signed-off-by: Arun C Ilango <arun.ilango@intel.com>

Commit:
prov/verbs: move ACK-ing rdmacm event inside the "process event" function

This is just a refactoring to simplify and prepare for a follow-up commit that
fixes a deadlock.

Signed-off-by: Arun C Ilango <arun.ilango@intel.com>

Commit:
prov/verbs: avoid a deadlock on multi-threaded access to EQ and EP

When one thread (1) closes an endpoint and another thread (2) reads an EQ that's
bound to the endpoint, a deadlock could happen in such a case:

1. Thread 2 reads an event from rdmacm event channel and is about to acquire
   eq->lock.
2. Thread 1 acquires eq->lock and calls rdma_destroy_id
3. Thread 1 hangs in rdma_destroy_id since the id has un-acknowledged events
4. Thread 1 hasn't acknolwedged the rdmacm event yet as it's stuck on acquiring
   eq->lock.

Fix the deadlock by expanding the scope of eq->lock when reading EQ. Now, a thread
reading the EQ will hold eq->lock until the rdmamcm event is acknowledged.

Signed-off-by: Arun C Ilango <arun.ilango@intel.com>

aingerson committed:
prov/shm: save completion address properly

The previous implementation was not saving the source address
properly if requested, either improperly casting the fi_addr_t
to a pointer or not passing in the correct address.

Change the completion function to take in a fi_addr_t instead of
void * and make sure it's passing in the correct value.

Signed-off-by: aingerson <alexia.ingerson@intel.com>

swelch committed:
prov/verbs: Remove XRC shared INI QP management lock

With the evolution of the EQ lock to make connection
processing thread-safe, the INI QP management lock
is no longer needed, the structure is protected by
the EQ lock.

Signed-off-by: Steve Welch <swelch@systemfabricworks.com>

nikhilnanal committed:
review-rework jenkins-pr  PR15
changed fabric labels and stage names to literals.
removed build_shmem.sh from tree
added commented tests back

Signed-off-by: nikhilnanal <nikhilnanal1@gmail.com>

Commit:
Merge pull request #5306 from a-ilango/rxm_fixes

prov/verbs, rxm: fix for a deadlock, refactoring & add env variable to force auto-progress

arn314 commented on Jenkinsfile:
```diff
@@ -0,0 +1,263 @@
+    environment {
+    // variables are referenced as ${env.VarName} in shell script.
+        ofi_install_dir="/home/build/ofi-Install/libfabric"
+    fabtests_install_dir="/home/build/ofi-Install/libfabric-fabtests"
```
nit: remove camel case (ofi-Install)

arn314 commented on Jenkinsfile:
```diff
@@ -0,0 +1,263 @@
+             }
+        }
+        
+        stage ('build') {
```
Can rename this as build-libfabric.

arn314 commented on Jenkinsfile:
```diff
@@ -0,0 +1,263 @@
+                sh """
+                  python3.7 contrib/intel/jenkins/build.py --builditem='libfabric'
+                  python3.7 contrib/intel/jenkins/build.py --builditem='libfabric' --build_mode='dbg'
+                  python3.7 contrib/intel/jenkins/build.py --builditem='libfabric' --build_mode='dl'
```
Rename build_mode to ofi_build_mode or you could do something like --libfabric_dl=yes, --ofi_debug=yes, making those options boolean. I think that better reflects the actual way it's used.

arn314 replied:
This can come as a later improvement if it's too much work to change.

arn314 commented on Jenkinsfile:
```diff
@@ -0,0 +1,263 @@
+    
+    stage('build IMPI_bm') {
+        steps {
+          withEnv(['PATH+EXTRA=/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/bin']) {
```
Can we use a variable here since it's getting repeated?

arn314 commented on Jenkinsfile:
```diff
@@ -0,0 +1,263 @@
+                        {
+                          sh """
+                            env
+                            ${env.ofi_install_dir}/${env.BRANCH_NAME}/${env.BUILD_NUMBER}/reg/bin/fi_info -f ${env.FABRIC}
```
` ${env.ofi_install_dir}/${env.BRANCH_NAME}/${env.BUILD_NUMBER}` seems to be repeated in multiple places. Can it be replaced by a variable?

nikhilnanal replied:
yes I have actually added code to push this into the runtest.py. so this statement will not be there. This is to be replaced by something like 
python3.7 runtest.py --basic_test=fi_info

arn314 commented on Jenkinsfile:
```diff
@@ -0,0 +1,263 @@
+       
+                 }
+                 
+                 stage('mlx-test-dl') {
```
This needs to be replaced by generic fabric names.

arn314 commented on contrib/intel/jenkins/build.py:
```diff
@@ -0,0 +1,193 @@
+parser.add_argument("--build_mode", help="select buildmode debug or dl", choices=['dbg','dl'])
+parser.add_argument("--other_benchmarks", help="build other non mpi benchmarks",
+                           choices=['shmem'])
+
```
The above code can be within a `if __name__ == "__main__":` block to allow for importing this file as a module in the future (if need arises).

arn314 commented on contrib/intel/jenkins/build.py:
```diff
@@ -0,0 +1,193 @@
+    build_mode = args.build_mode
+else:
+    build_mode = 'reg'
+
```
ditto

arn314 commented on contrib/intel/jenkins/build.py:
```diff
@@ -0,0 +1,193 @@
+                          '--enable-usnic=no', '--enable-psm=no', '--enable-psm2=yes',
+                          '--enable-verbs=yes', '--enable-rxd=yes', '--enable-rxm=yes', 
+                          '--enable-sockets=yes', '--enable-tcp=yes', '--enable-udp=yes',
+                          '--enable-rxd=yes', '--enable-shm=yes', '--with-psm2-src={}/opa-psm2-lib'.format(workspace)]
```
Use common variables instead of repeating code below.

arn314 commented on contrib/intel/jenkins/build.py:
```diff
@@ -0,0 +1,193 @@
+                '--with-libfabric={}'.format(install_path)] 
+    else:
+        config_cmd = ['./configure', '--prefix={}'.format(install_path),
+                '--with-libfabric={}'.format(install_path)]
```
ditto. Just append "--enable-debug" to the list if mode == 'dbg'

arn314 commented on contrib/intel/jenkins/build.py:
```diff
@@ -0,0 +1,193 @@
+    if (mpi == 'ompi'):
+        mpistr = '-mpi-'
+    elif (mpi == 'mpich'):
+        mpistr= '-'
```
why no string for mpich?

nikhilnanal replied:
the flag for mpich is --enable-fortran while for ompi it is --enable-mpi-fortran

arn314 commented on contrib/intel/jenkins/build.py:
```diff
@@ -0,0 +1,195 @@
+        os.environ['LD_LIBRARY_PATH'] = ""
+        mpicc_path = "{}/bin/mpicc".format(mpi_install_path)
+
+    cmd=" ".join([mpicc_path, '-lz', "{}/mpi_stress/mpi_stress.c" \
```
tip: you could avoid the backslash escapes as long as you are breaking a line enclosed within `(` and `)`.

nikhilnanal replied:
yeah I think I just added it for consistency sake. 
so that I don't miss it somewhere else.

arn314 commented on Jenkinsfile:
```diff
@@ -0,0 +1,252 @@
+                            python3.7 runtests.py --prov=tcp --build_mode='dl'
+                            python3.7 runtests.py --prov=udp --build_mode='dl'
+                            python3.7 runtests.py --prov=sockets --build_mode='dl'               
+                            cd ${env.WORKSPACE}  
```
tip: if you enclose the shell block within `(` and `)` (a subshell) and `cd` to a different dir within the subshell, you don't need to `cd` back into original directory.

nikhilnanal replied:
I ll try this, just to be sure that there are no jenkinsfile quirks to this

arn314 commented on contrib/intel/jenkins/runtests.py:
```diff
@@ -0,0 +1,64 @@
+parser.add_argument("--build_mode", help="specify the build configuration", \
+                     choices = ["dbg", "dl"])
+parser.add_argument("--fi_info_test", help="run the fi_info test", \
+                    choices=['yes','no'])
```
Why is this a separate option?

nikhilnanal replied:
because there is no option for testname, as all tests except this one are run with the provider option specified and fi_info does not take provider argument.

arn314 replied:
Hmm. fi_info should take a provider argument.

arn314 commented on contrib/intel/jenkins/run.py:
```diff
@@ -0,0 +1,83 @@
+#run fi_info test
+def fi_info_test(mode):
+    cmdpath = "{}/{}/{}/{}/bin/".format(ci_site_config.install_dir, \
+               brname, bno, mode)
```
tip: you may be able to use the new Python "fstrings" to make the formatting more readable.

arn314 commented on contrib/intel/jenkins/tests.py:
```diff
@@ -0,0 +1,309 @@
+        assert(self.osu_mpi_path)
+        for root, dirs, files in os.walk(self.osu_mpi_path):
+            for file in files:
+                if file in self.two_proc_test_pattern:
```
Rename file to test or benchmark. Rename `two_proc_test_pattern` to `two_proc_tests` and make it a dict.

arn314 commented on contrib/intel/jenkins/tests.py:
```diff
@@ -0,0 +1,309 @@
+                         mpitype,  hosts, ofi_build_mode, util_prov)
+        
+         
+        if((self.core_prov == "verbs" or self.core_prov =="psm2")):
```
nit: add space b/w if and the open parenthesis.

arn314 replied:
Add comments for this condition as well.

arn314 commented on contrib/intel/jenkins/tests.py:
```diff
@@ -0,0 +1,309 @@
+    @property
+    def execute_condn(self):
+        return True if (self.mpi != 'ompi' or \
+                        self.ofi_build_mode != 'dbg') else  False
```
Add comments for these conditions.

arn314 commented on contrib/intel/jenkins/build.py:
```diff
@@ -0,0 +1,190 @@
+    if (mpi == 'ompi'):
+        cmd.append("--enable-mpi-fortran=no")
+    elif (mpi == 'mpich'):
+        cmd.append("--enable-fortran=no")
```
This is much better.

Commit:
Merge pull request #5327 from swelch/rm_ini_mgmt_lock

prov/verbs: Remove XRC shared INI QP management lock

arn314 commented on contrib/intel/jenkins/tests.py:
```diff
@@ -0,0 +1,309 @@
+        self.libfab_installpath = "{}/{}/{}/{}".format(ci_site_config.install_dir,
+                                  self.branchname, self.buildno, self.ofi_build_mode)
+ 
+        self.env = [("FI_VERBS_MR_CACHE_ENABLE", "1"),\
```
This env variable has been deprecated from v1.8 onwards. If you plan to back-port Jenkinsfile to the v1.7.x, then you can add this there.

nikhilnanal replied:
aren't we planning to build/test against all branches right.

nikhilnanal replied:
wouldn't it be better to retain it than remove it. it may be harder to trace back the changes required later in case back porting is required? whereas at the moment it will work for both the cases.

arn314 commented on contrib/intel/jenkins/tests.py:
```diff
@@ -0,0 +1,309 @@
+        if os.path.isfile(efile):
+            return efile
+        elif os.path.isfile(efile_old):
+            return efile_old
```
The efile_old is not needed in the latest versions. I think it's needed for v1.6 or older. This change will be needed there if you ever backport Jenkinsfile to v1.6

nikhilnanal replied:
wouldn't it be better to retain it than remove it. it may be harder to trace back the changes required later in case back porting is required? whereas at the moment it will work for both the cases.

arn314 replied:
Yes, it will work for both cases if you retain it. It's just unnecessary code to maintain.

arn314 commented on contrib/intel/jenkins/tests.py:
```diff
@@ -0,0 +1,309 @@
+            opts = "{} -s {} ".format(opts, common.get_node_name(self.server, 
+                    self.nw_interface)) # include common.py
+            opts = "{} -c {} ".format(opts, common.get_node_name(self.client, 
+                    self.nw_interface)) # from common.py
```
Drop the comments.

arn314 commented on contrib/intel/jenkins/tests.py:
```diff
@@ -0,0 +1,309 @@
+
+        efile = self.get_exclude_file()
+        if efile:
+            opts = "{} -R ".format(opts)
```
-R option has been deprecated from v1.8 onwards.

arn314 commented:
Is it possible to change `continuous-integration/jenkins/pr-merge` to `Intel CI`?

jswaro committed:
core: Add support for traffic class selection

This commit introduces the capability to select a traffic class when
creating libfabric objects with the goal to allow applications to
utilize traffic shaping mechanisms exposed to providers.

In this proposal, the fi_domain, fi_tx_attr, and fi_rx_attributes are
modified to add the new fi_traffic_class structure to the structure definition.
The fi_traffic_class structure provides a 'type' field, indicating the type of
traffic class value that is represented in the 'value' field of the structure.
Two different types of traffic class are presented with this commit; FI_TC_TYPE_LABEL,
and FI_TC_TYPE_DSCP. Libfabric defines generalized traffic classes via community
defined labels and expected behavior for those labels, which fall under the category
of FI_TC_TYPE_LABEL. Providers may also define provider specific labels utilizing the
most significant 32 bits of the value field. To provide applications with greater
control over traffic class selection, this proposal exposes the ability to specify
a specify DSCP value to be associated with libfabric objects, using the value field set
to the DSCP value, and type set to FI_TC_TYPE_DSCP.

Additionally, to simplify the traffic class selection process, this commit introduces
a heirarchal inheritance model with the fi_traffic_class structure. When an fi_domain
is created, the application can choose to select a traffic class, or allow the provider
to select a default traffic class. The traffic class associated with the domain is used
as the default for all objects created within the domain. This allows an
application to request a general traffic class for most traffic flows, while
allowing the application to finely control traffic at the granularity of the tx context
or rx context.

The original proposal to OFIWG did not include the rx context modification. However,
multi-part data transfer operations such as rendezvous messaging benefit from the differentiation of
request and response traffic shaping flows. The traffic class field in the rx attribute is
only intended to apply to these multi-part data transfer operations and does not imply any
relation with the tx attribute of a peer endpoint.

Signed-off-by: James Swaro <jswaro@cray.com>
Signed-off-by: Sean Hefty <sean.hefty@intel.com>

aingerson committed:
prov/shm: workaround for CMA ptrace issue on Ubuntu

Ubuntu does not default to the corret pthread_scope for using CMA.
CMA needs a pthread_scope of 0 (no restrictions).

Check ptrace_scope settings and adjust max_msg_size to disable CMA path

Signed-off-by: aingerson <alexia.ingerson@intel.com>

aingerson committed:
fabtests/common: do not use size if it exceeds provider limit

There is currently no check for only testing message sizes within
the limits of the provider. Add the check to avoid expected failures.

Remove the check from the dgram_pingpong test which also calls the
updated size check function.

Signed-off-by: aingerson <alexia.ingerson@intel.com>

Commit:
Merge pull request #5329 from aingerson/master

prov/shm: save completion address properly

zhngaj committed:
Bump the protocol version from 2 to 3.

Bump version in preparation for upcoming changes to the EFA provider
protocol.

Signed-off-by: Jie Zhang <zhngaj@amazon.com>

rajachan committed:
Merge pull request #5320 from zhngaj/efa-shm-support

prov/efa: Enable shared memory based transfer with SHM provider for intra-node communication

ofiwg-bot committed:
Updated nroff-generated man pages

Signed-off-by: OFIWG Bot <ofiwg@lists.openfabrics.org>

ofiwg-bot committed:
Merge pull request #5332

More tasty nroff man pages for you, fresh out of the oven!

Commit:
Merge pull request #5330 from aingerson/shm

prov/shm: ptrace check and adjustment, fabtests/common: avoid testing unsupported sizes

rajachan committed:
prov/efa: Rework function signatures of generic tx functions

This commit changes the rather messy function signatures of
rxr_ep_tx_entry_init(), rxr_generic_tx_entry_init(), rxr_tx(), and
rxr_generic_rma() to use `struct fi_msg` or `struct fi_msg_rma` as
applicable. The send descriptor was also getting dropped in some paths,
which is now fixed by using the msg structures.

Signed-off-by: Raghu Raja <craghun@amazon.com>

arn314 commented on contrib/intel/jenkins/run.py:
```diff
@@ -0,0 +1,91 @@
+
+
+if __name__ == "__main__":
+    pass
```
can drop this if not being used.

arn314 commented on contrib/intel/jenkins/runtests.py:
```diff
@@ -0,0 +1,62 @@
+    for host in ci_site_config.node_map[node]:
+        hosts.append(host)
+
+    for prov in common.prov_list:
```
Do we need to iterate through prov_list? Would just checking if args_prov is in the list be enough?

nikhilnanal replied:
we need to run this code for every core and util combination. there could be multiple matches for a core provider since the list contains tuples of core+util combinations, but a single if  staement would just stop at the first match.

arn314 reviewed (approved):
Looks good. Improvements can be made but I guess this is good enough for now.

rajachan committed:
Merge pull request #5338 from rajachan/tx-fnsig

prov/efa: Rework function signatures of generic tx functions

Commit:
Merge pull request #5334 from shefty/master

core: Add support for traffic class selection

ofiwg-bot committed:
Updated nroff-generated man pages

Signed-off-by: OFIWG Bot <ofiwg@lists.openfabrics.org>

ofiwg-bot committed:
Merge pull request #5343

More tasty nroff man pages for you, fresh out of the oven!

Commit:
Merge pull request #5261 from AMcKinley15/test_runner

Multinode tests in runfabtests.sh
